README.md


[![npm version](https://img.shields.io/npm/v/neoui-kero.svg)](https://www.npmjs.com/package/neoui-kero)
[![Build Status](https://img.shields.io/travis/iuap-design/neoui-kero/master.svg)](https://travis-ci.org/iuap-design/neoui-kero)
[![devDependency Status](https://img.shields.io/david/dev/iuap-design/neoui-kero.svg)](https://david-dm.org/iuap-design/neoui-kero#info=devDependencies)
[![NPM downloads](http://img.shields.io/npm/dm/neoui-kero.svg?style=flat)](https://npmjs.org/package/neoui-kero)

[中文文档](./README_CN.md)


##  Introduction
[neoui-kero](http://tinper.org/dist/kero/docs/module.html) A adapter for [kero](http://tinper.org/dist/kero/index.html) and [tinper neoui](http://tinper.org/dist/neoui/index.html)。


## Features


* Create component
*  Simplify data manipulation

## Quickstart

###Get neoui-kero

* npm
```
    npm install neoui-kero
```

* cdn
```
    //design.yonyoucloud.com/static/neoui-kero/latest/neoui-kero.js
```
### Introducing  neoui-kero
- ES6
```
    <!-- neoui-kero -->
    import { u } from "neoui-kero"

```
- HTML

```
    <!-- neoui-kero -->
    <script src="//design.yonyoucloud.com/static/neoui-kero/latest/neoui-kero.js"></script>
```
**注**： neoui-kero often used in conjunction with [neoui](https://github.com/iuap-design/tinper-neoui), [kero](https://github.com/iuap-design/kero)


### Use


```

<input id="demo_input" u-meta='{"id":"t1","type":"string","data":"dt1","field":"f1"}' />
<div id="demo_div">&lt;/div></code></pre>
</div>
```

```
// JS
var app,viewModel;

viewModel = {
    dt1: new u.DataTable({
        meta:{
            f1:{
                type:'string'
            }
        }
    })
};


app = u.createApp({
    el:'body',
    model:viewModel
});

var r = viewModel.dt1.createEmptyRow();
r.setValue('f1','Hello World');



var demoInput = document.getElementById('demo_input');
var demoDiv = document.getElementById('demo_div');

var getDtValue = function() {
    var dtVal = viewModel.dt1.getValue('f1');
    demoDiv.innerHTML = dtVal;
};
demoInput.addEventListener('blur',getDtValue);
getDtValue();

```

Read the [Develop documentation](http://tinper.org/dist/kero/docs/module.html)


## Contributing

### Develop

Developers can participate in the development of neoui-kero,  but also can be based on neoui-kero two development


neoui-kero use gulp.js and webpack build the project.


clone：

```
$ git clone git@github.com:iuap-design/neoui-kero.git
```

install：

```
$ npm install
```

build：

```
$ npm run product
```


### Feedback

If you encounter any problems , submit [issues](https://github.com/iuap-design/neoui-kero/issues),or pull request。

[PR code](CONTRIBUTING.md)


## Licence 版权

[MIT](./LICENSE)


CHANGELOG.md
<a name="3.2.4"></a>
## [3.2.4](https://github.com/iuap-design/neoui-kero/compare/v3.2.3...v3.2.4) (2017-08-03)


### Bug Fixes

* 修改booleanRender在grid的edittable设置为false时，还能编辑的bug ([d8762b7](https://github.com/iuap-design/neoui-kero/commit/d8762b7))
* 修改年月控件设置format不起作用的bug ([7759858](https://github.com/iuap-design/neoui-kero/commit/7759858))
* 将grid中的下拉框ui风格与表单中的下拉框风格保持统一 ([efcc911](https://github.com/iuap-design/neoui-kero/commit/efcc911))
* 解决float类型输入框不能输入小键盘减号bug ([a997e5a](https://github.com/iuap-design/neoui-kero/commit/a997e5a))


### Features

* 增加switchRender可以赋值0和1来设置开关 ([6654ccf](https://github.com/iuap-design/neoui-kero/commit/6654ccf))
* 当在多选的时候，设置selectChangeDatatable为true时，选中一个数据就会动态的去改变datatable ([add000e](https://github.com/iuap-design/neoui-kero/commit/add000e))
* 支持radio控件的boolean类型的true和false与true和false比较 ([9e74018](https://github.com/iuap-design/neoui-kero/commit/9e74018))



<a name="3.2.3"></a>
## [3.2.3](https://github.com/iuap-design/neoui-kero/compare/v3.2.2...v3.2.3) (2017-06-22)


### Bug Fixes

* 解决数字组件校验失效的问题 ([1f6afdc](https://github.com/iuap-design/neoui-kero/commit/1f6afdc))


### Features

* grid支持自动扩展宽度的render为autoWidthRender，支持textArea编辑 ([3fdf9c6](https://github.com/iuap-design/neoui-kero/commit/3fdf9c6))



<a name="3.2.2"></a>
## [3.2.2](https://github.com/iuap-design/neoui-kero/compare/v3.2.1...v3.2.2) (2017-05-23)


### Bug Fixes

* 解决树清空选中后原有选中行依旧显示为选中的问题 ([c49fa1b](https://github.com/iuap-design/neoui-kero/commit/c49fa1b))


### Features

* 新增Cascader组件 ([027aaef](https://github.com/iuap-design/neoui-kero/commit/027aaef))



<a name="3.2.1"></a>
## [3.2.1](https://github.com/iuap-design/neoui-kero/compare/v3.2.0...v3.2.1) (2017-04-19)


### Bug Fixes

* 单选、下拉、多选getShowValue获取不到值 ([860f4c1](https://github.com/iuap-design/neoui-kero/commit/860f4c1))
* 解决clockpickler组件第一次选择00:00无法存储数据的问题 ([0df9705](https://github.com/iuap-design/neoui-kero/commit/0df9705))
* 解决tree通过datatable删除数据时报错的问题 ([12c2900](https://github.com/iuap-design/neoui-kero/commit/12c2900))



<a name="3.2.0"></a>
# [3.2.0](https://github.com/iuap-design/neoui-kero/compare/v3.1.27...v3.2.0) (2017-03-24)


### Bug Fixes

* 产出重命名 ([89689f2](https://github.com/iuap-design/neoui-kero/commit/89689f2))
* 币种组件不再显示币种效果 ([70dac96](https://github.com/iuap-design/neoui-kero/commit/70dac96))


### Features

* grid组件使用comboRender及radioRender时可以直接在column上定义datasource ([2cbbe2a](https://github.com/iuap-design/neoui-kero/commit/2cbbe2a))
* neoui-kero支持按照组件产出对应的js文件以方便项目使用 ([f0bc2b9](https://github.com/iuap-design/neoui-kero/commit/f0bc2b9))
* 输入组件支持placeholder属性 ([9d67ed0](https://github.com/iuap-design/neoui-kero/commit/9d67ed0))



<a name="3.1.27"></a>
## [3.1.27](https://github.com/iuap-design/neoui-kero/compare/v3.1.26...v3.1.27) (2017-02-23)



<a name="3.1.26"></a>
## [3.1.26](https://github.com/iuap-design/neoui-kero/compare/v3.1.25...v3.1.26) (2017-02-20)


### Features

* adapter层setValue增加beforeSetValue调用，方便组件个性化设置 ([44706df](https://github.com/iuap-design/neoui-kero/commit/44706df))
* 对框架进行升级，新增源码仓库compox、compox-util、kero-fetch、neoui-kero、neoui-kero-mixin ([880cb40](https://github.com/iuap-design/neoui-kero/commit/880cb40))



<a name="3.1.25"></a>
## [3.1.25](https://github.com/iuap-design/neoui-kero/compare/v3.1.22...v3.1.25) (2017-02-10)


### Bug Fixes

* editType 默认值 ([f535189](https://github.com/iuap-design/neoui-kero/commit/f535189))
* grid中字段为float时编辑问题、月日控件多次赋值问题、tree中onclick问题 ([297dbc8](https://github.com/iuap-design/neoui-kero/commit/297dbc8))
* time设置值 ([7aefd7c](https://github.com/iuap-design/neoui-kero/commit/7aefd7c))


### Features

* 架构升级，将原有仓库中的部分功能拆分至四个新的仓库中 ([6f942f6](https://github.com/iuap-design/neoui-kero/commit/6f942f6))



<a name="3.1.22"></a>
## [3.1.22](https://github.com/iuap-design/neoui-kero/compare/v3.1.21...v3.1.22) (2017-01-10)



<a name="3.1.19"></a>
## [3.1.19](https://github.com/iuap-design/neoui-kero/compare/v3.1.18...v3.1.19) (2016-12-28)


### Bug Fixes

* 分页 ([b74aa28](https://github.com/iuap-design/neoui-kero/commit/b74aa28))
* 多语组件与datatable关联打通 ([a6a8e9a](https://github.com/iuap-design/neoui-kero/commit/a6a8e9a))



<a name="3.1.18"></a>
## [3.1.18](https://github.com/iuap-design/neoui-kero/compare/v3.1.17...v3.1.18) (2016-12-22)


### Bug Fixes

* changelog样式 ([7079bde](https://github.com/iuap-design/neoui-kero/commit/7079bde))
* 云表单radio、单选框其他输入框disable设置 ([3717e56](https://github.com/iuap-design/neoui-kero/commit/3717e56))



<a name="3.1.17"></a>
## [3.1.17](https://github.com/iuap-design/neoui-kero/compare/v3.1.16...v3.1.17) (2016-12-14)


### Bug Fixes

* form情况下按照showFix展示 ([986f72f](https://github.com/iuap-design/neoui-kero/commit/986f72f))
* i18n & add json files & test example ([8f7a9e3](https://github.com/iuap-design/neoui-kero/commit/8f7a9e3))
* i18n demo ([fa27546](https://github.com/iuap-design/neoui-kero/commit/fa27546))



<a name="3.1.16"></a>
## [3.1.16](https://github.com/iuap-design/neoui-kero/compare/v3.1.15...v3.1.16) (2016-12-07)


### Bug Fixes

* grid中checkbox值变时dataTable字段值不变 ([12011c8](https://github.com/iuap-design/neoui-kero/commit/12011c8))
* 表格中float类型在行之间切换编辑数据会错乱 ([7c302ad](https://github.com/iuap-design/neoui-kero/commit/7c302ad))



<a name="3.1.15"></a>
## [3.1.15](https://github.com/iuap-design/neoui-kero/compare/v3.1.14...v3.1.15) (2016-11-30)


### Bug Fixes

* 优化ztree选中之后更新至datatable会出现错误的情况 ([deb7f3f](https://github.com/iuap-design/neoui-kero/commit/deb7f3f))



<a name="3.1.14"></a>
## [3.1.14](https://github.com/iuap-design/neoui-kero/compare/v3.1.13...v3.1.14) (2016-11-25)


### Bug Fixes

* 翻页及pageSize改变时清空grid数据。datatable发生改变时设置silence为true。 ([cae6110](https://github.com/iuap-design/neoui-kero/commit/cae6110))



<a name="3.1.13"></a>
## [3.1.13](https://github.com/iuap-design/neoui-kero/compare/v3.1.12...v3.1.13) (2016-11-23)


### Bug Fixes

* 解决tree传递数据的时候除了id pid name，其他字段丢失的情况 ([db11473](https://github.com/iuap-design/neoui-kero/commit/db11473))



<a name="3.1.12"></a>

js/tree.js
/**
 * Module : Kero tree adapter
 * Author : Kvkens(yueming@yonyou.com)
 * Date	  : 2016-08-16 10:44:14
 */
import {BaseAdapter} from './baseAdapter';
import {Year} from 'neoui/js/neoui-year';
import {getJSObject,getFunction} from 'neoui-sparrow/js/util';
import {DataTable} from 'kero/js/dataTable/indexDataTable';
import {compMgr} from 'neoui-sparrow/js/compMgr';

var TreeAdapter = BaseAdapter.extend({
	
	initialize: function(options) {
		var opt = options['options'] || {},
				viewModel = options['model'];
		var element = typeof options['el'] === 'string' ? document.querySelector(options['el']) : options['el'];
		var app = options['app'];
		this.id = opt['id'];
		options = opt;
		
		var oThis = this;
		this.dataTable = getJSObject(viewModel, options["data"]);
		this.element = element;
		this.$element = $(element);
		this.id = options['id'];
		this.element.id = this.id;
		this.options = options;
		this.events = $.extend(true, {}, options.events);
		var treeSettingDefault = {
			//			async: {  //缓加载
			//				enable: oThis.options.asyncFlag,
			//				url: oThis.options.asyncFun
			//			},
			data: {
				simpleData: {
					enable: true
				}
			},
			check: {
				chkboxType: {
					"Y": "",
					"N": ""
				}
			},
			callback: {
				//点击前
				beforeClick: function(e, id, node) {
					if (oThis.events.beforeClick) {
						getFunction(viewModel, oThis.events.beforeClick)(e, id, node);
					}
				},
				// 选中/取消选中事件
				onCheck: function(e, id, node) {

					var nodes= oThis.tree.getChangeCheckedNodes();
					for(var i=0 ;i<nodes.length;i++){
						// 获取到节点的idValue
						var idValue=nodes[i].id;
						// 根据idValue查找到对应数据的rowId
						var rowId = oThis.getRowIdByIdValue(idValue);
						var index = oThis.dataTable.getIndexByRowId(rowId);
						if (nodes[i].checked) {
							// 选中数据行
							nodes[i].checkedOld = true;
							if (oThis.tree.setting.view.selectedMulti == true) {
								oThis.dataTable.addRowsSelect([index]);
							} else {
								oThis.dataTable.setRowSelect(index);
							}
						} else {
							nodes[i].checkedOld = false;
							oThis.dataTable.setRowUnSelect(index);
						}
					}
				},
				// 单选时点击触发选中
				onClick: function(e, id, node) {
					//点击时取消所有超链接效果
					$('#'+id+' li').removeClass('focusNode');
					$('#'+id+' a').removeClass('focusNode');
					//添加focusNode样式
					$('#'+node.tId).addClass('focusNode');
					$('#'+node.tId+'_a').addClass('focusNode');
					// 获取到节点的idValue
					var idValue = node.id;
					// 根据idValue查找到对应数据的rowId
					var rowId = oThis.getRowIdByIdValue(idValue);
					var index = oThis.dataTable.getIndexByRowId(rowId);
					oThis.dataTable.setRowSelect(index);
					if (oThis.events.onClick) {
						getFunction(viewModel, oThis.events.onClick)(e, id, node);
					}
				}
			}

		};

		var setting = {};
		if (this.options.setting) {
			//if (typeof(JSON) == "undefined")
			//	setting = eval("(" + this.options.setting + ")");
			//else
				setting = getJSObject(viewModel, this.options.setting) || getJSObject(window, this.options.setting);
		}

		// 遍历callback先执行默认之后再执行用户自定义的。
		var callbackObj = treeSettingDefault.callback;
		var userCallbackObj = setting.callback;
		for(var f in callbackObj){
			var fun = callbackObj[f],
				userFun = userCallbackObj && userCallbackObj[f];
			if(userFun){
				var newF = function(){
					fun.apply(this,arguments);
					userFun.apply(this,arguments);
				}
				userCallbackObj[f] = newF;
			}
		}


		var treeSetting = $.extend(true,{}, treeSettingDefault, setting);

		var treeData = [];
		// 根据idField、pidField、nameField构建ztree所需data
		var data = this.dataTable.rows();
		if (data.length > 0) {
			if (this.options.codeTree) {
				// 首先按照string进行排序
				data.sort(function(a, b) {
					var aObj = a.data;
					var bObj = b.data;
					var v1 = aObj[oThis.options.idField].value + '';
					var v2 = bObj[oThis.options.idField].value + '';
					try {
						return v1.localeCompare(v2);
					} catch (e) {
						return 0;
					}
				});
				var idArr = new Array();
				$.each(data, function() {
					var dataObj = this.data;
					var idValue = dataObj[oThis.options.idField].value;
					idArr.push(idValue);
				});
				var preValue = '';
				$.each(data, function() {
					var value = {};
					var dataObj = this.data;
					var idValue = dataObj[oThis.options.idField].value;
					var nameValue = dataObj[oThis.options.nameField].value;
					var pidValue = '';
					var startFlag = -1;
					// 如果当前值包含上一个值则上一个值为pid
					if (preValue != '') {
						var startFlag = idValue.indexOf(preValue);
					}
					if (startFlag == 0) {
						pidValue = preValue;
					} else {
						for (var i = 1; i < preValue.length; i++) {
							var s = preValue.substr(0, i);
							var f = idValue.indexOf(s);
							if (f == 0) {
								var index = $.inArray(s, idArr);
								if (index > 0 || index == 0) {
									pidValue = s;
								}
							} else {
								break;
							}
						}

					}
					value['id'] = idValue;
					value['pId'] = pidValue;
					value['name'] = nameValue;

					treeData.push(value);
					preValue = idValue;
				});
			} else {
				var values = new Array();
				$.each(data, function() {
					var value = {};
					var dataObj = this.data;
					var idValue = dataObj[oThis.options.idField].value;
					var pidValue = dataObj[oThis.options.pidField].value;
					var nameValue = dataObj[oThis.options.nameField].value;

					value['id'] = idValue;
					value['pId'] = pidValue;
					value['name'] = nameValue;
					treeData.push(value);
				});
			}
		}

		this.tree = $.fn.zTree.init(this.$element, treeSetting, treeData);


		// dataTable事件
		this.dataTable.on(DataTable.ON_ROW_SELECT, function(event) {
			/*index转化为grid的index*/
			$.each(event.rowIds, function() {
				var row = oThis.dataTable.getRowByRowId(this);
				var dataObj = row.data;
				var idValue = dataObj[oThis.options.idField].value;
				var node = oThis.tree.getNodeByParam('id', idValue);
				if (oThis.tree.setting.view.selectedMulti == true  && !node.checked) {
					oThis.tree.checkNode(node,true,false,true);
				} else {
					oThis.tree.selectNode(node, false);
				}
			});
		});

		this.dataTable.on(DataTable.ON_ROW_UNSELECT, function(event) {
			/*index转化为grid的index*/
			$.each(event.rowIds, function() {
				var row = oThis.dataTable.getRowByRowId(this);
				var dataObj = row.data;
				var idValue = dataObj[oThis.options.idField].value;
				var node = oThis.tree.getNodeByParam('id', idValue);
				if (oThis.tree.setting.view.selectedMulti == true && node.checked) {
					oThis.tree.checkNode(node,false,true,true);
				} else {
					oThis.tree.cancelSelectedNode(node)
				}
			});
		});

		this.dataTable.on(DataTable.ON_INSERT, function(event) {
			//var gridRows = new Array();
			var dataArray=[],nodes=[];
			var hasChild=false;//是否含有子节点
			$.each(event.rows, function() {
				var value = {},hasPar=false;
				var dataObj = this.data;
				var idValue = dataObj[oThis.options.idField].value;
				var pidValue = dataObj[oThis.options.pidField].value;
				var nameValue = dataObj[oThis.options.nameField].value;
				value['id'] = idValue;
				value['pId'] = pidValue;
				value['name'] = nameValue;
				var childNode=oThis.tree.getNodeByParam('pid', idValue);
				var pNode = oThis.tree.getNodeByParam('id', pidValue);
				if(childNode&&childNode.length>0){
					hasChild=true;
				}
				if(pNode&&pNode.length>0){
					hasPar=true;
					//oThis.tree.addNodes(pNode, value, true);
				}
				if(!hasChild&&hasPar){//不存在子节点,存在父节点之间插入
					oThis.tree.addNodes(pNode, value, true);
				}else{
					dataArray.push(value);
				}
				
			})
			if(!hasChild){//如果没有子节点，将当前节点作为根节点之间插入
				nodes=oThis.tree.transformTozTreeNodes(dataArray);
				oThis.tree.addNodes(null, nodes, true);
			}else{//如果含有子节点,重新渲染
				
			}
		});


		this.dataTable.on(DataTable.ON_DELETE, function(event) {
			/*index转化为grid的index*/
			var gridIndexs = new Array();
			if (this.deleteRows.length > 0) {

				for (var i = 0; i < this.deleteRows.length; i++) {
					var row = this.deleteRows[i];
					var dataObj = row.data;
					var idValue = dataObj[oThis.options.idField].value;
					var node = oThis.tree.getNodeByParam('id', idValue);
					oThis.tree.removeNode(node)
				}

			}
		});

		this.dataTable.on(DataTable.ON_DELETE_ALL, function(event) {
			var nodes = oThis.tree.getNodes();
			for (var i = 0, l = nodes.length; i < l; i++) {
				var node = oThis.tree.getNodeByParam('id', nodes[i].id);
				oThis.tree.removeNode(node);
				i--;
				l = nodes.length;
			}
		});

		// 加载数据,只考虑viewModel传入grid
		this.dataTable.on(DataTable.ON_LOAD, function(data) {
			var data = oThis.dataTable.rows();
			if (data.length > 0) {
				var values = new Array();
				$.each(data, function() {
					var value = {};
					var dataObj = this.data;
					var idValue = dataObj[oThis.options.idField].value;
					var pidValue = dataObj[oThis.options.pidField].value;
					var nameValue = dataObj[oThis.options.nameField].value;

					value['id'] = idValue;
					value['pId'] = pidValue;
					value['name'] = nameValue;
					treeData.push(value);
				});
			}

			this.tree = $.fn.zTree.init(this.$element, treeSetting, treeData);
		});

		this.dataTable.on(DataTable.ON_VALUE_CHANGE, function(event) {
			var row = oThis.dataTable.getRowByRowId(event.rowId);
			if (!row) return
			var treeArray=oThis.tree.getNodes();
			var id = row.getValue(oThis.options.idField)
			var node = oThis.tree.getNodeByParam('id', id);
			if(!node&&treeArray){//如果node为null则取树数组的最后一个节点

				node=treeArray[treeArray.length-1];
			}
			var field = event.field;
			var value = event.newValue;
			if(oThis.options.idField==field&&node){
				node.id=value;
				oThis.tree.updateNode(node);
			}
			if (oThis.options.nameField == field&&node){
				node.name = value
				oThis.tree.updateNode(node)
			}
			else if (oThis.options.pidField == field){
				var targetNode = oThis.tree.getNodeByParam('id', value);
				oThis.tree.moveNode(targetNode, node,"inner")				
			}
		});	

		// 通过树id获取dataTable的rowId
		this.getRowIdByIdValue = function(idValue) {
			var oThis = this;
			var rowId = null;
			$.each(this.dataTable.rows(), function() {
				var dataObj = this.data;
				var id = this.rowId;
				if (dataObj[oThis.options.idField].value == idValue) {
					rowId = id;
				}
			})
			return rowId;
		}

		return this;
	},

	getName : function() {
		return 'tree'
	}


});

compMgr.addDataAdapter({
	adapter: TreeAdapter,
	name: 'tree'
		//dataType: 'float'
});

export {TreeAdapter};
